[vcpkg] Use triplet file results in dependency logic expressions

Rastaban opened the pull request:
Previously the logic expression evaluation would do a sub string search of the triplet name to determine if an identifier was 'true' or 'false'.  With this PR it uses the actual values set inside the triplet file.  Most of the changes to vcpkg.exe was refactoring the dependency logic to allow running the triplet file earlier so its results could be used in the qualified dependency resolution.

This PR also fixes the previously undocumented "Supports" CONTROL tag to also use the logic expression evaluation.  Currently the only place the supports tag is used is in the `ci` command.  This will help clean up the results in the CI pipeline - especially when things switch over to using the checked in baseline.  The tag could also be used to provide better error messages on failure, but I will leave that for someone else to do in a later PR.

The unit tests are not complete.

Rastaban committed:
remove unfinished "supports" tag

Rastaban committed:
extract "supports" from control files
But do nothing with the value

Rastaban committed:
Start `Supports` documentation

Rastaban committed:
Use Supports in a bunch of control files

I only tried matching the already existing logic in the portfile.cmake.

cbezault committed:
Cmake var provider (#8)

cbezault committed:
Cmake var provider (#9)

cbezault committed:
fix windows build (#10)

Rastaban committed:
Add missing files to build

Rastaban committed:
Merge branch 'master' into cleanup_supports

cbezault committed:
Fix test (#11)

Rastaban committed:
adding hooks for cmake variables in expressions

Rastaban committed:
Adding hooks for 'supports' in CI test

cbezault committed:
Fix test (#12)

Rastaban committed:
Merge branch 'master' into cleanup_supports

Rastaban committed:
Add overrides to evaluation environment

Rastaban committed:
use "supported" tag in CI testing

Rastaban committed:
cleanup comment

vicroms commented:
There's a bug in the dependency graph calculation, transitive dependencies are not being added to the install plan.

Examples:

```cmake
PS D:\src\vcpkg> ./vcpkg install libzippp
The following packages will be built and installed:
  * libzip[core,bzip2,openssl]:x86-windows
    libzippp[core]:x86-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/2: libzip:x86-windows

# Missing bzip and openssl packages in install plan
```

```cmake
PS D:\src\vcpkg> ./vcpkg install opencv
The following packages will be built and installed:
    opencv[dnn,jpeg,webp,opengl,tiff,png,core]:x86-windows
  * opencv4[core,dnn,jpeg,webp,opengl,tiff,png]:x86-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/2: opencv4:x86-windows

# Missing a lot of dependencies
```

vicroms commented:
It seems to be ignoring `Build-Depends:` completely.
E.g.:

`ports/libzip/CONTROL`
```
Source: libzip
Version: rel-1-5-2
Homepage: https://github.com/nih-at/libzip
Build-Depends: zlib
Default-Features: openssl, bzip2
Description: A library for reading, creating, and modifying zip archives.

Feature: bzip2
Build-Depends: bzip2
Description: Support bzip2-compressed zip archives

Feature: openssl
Build-Depends: openssl
Description: AES (encryption) support using OpenSSL
```

```
PS D:\src\viromer\vcpkg> ./vcpkg install libzip
The following packages will be built and installed:
    libzip[openssl,core,bzip2]:x86-windows
Starting package 1/1: libzip:x86-windows
```

Ignored direct dependency on `zlib` and dependencies from features `bzip2` and `openssl` (which in turn depends on `openssl-windows`).

cbezault commented on toolsrc/src/vcpkg/build.cpp:
```diff
@@ -62,10 +71,31 @@ namespace vcpkg::Build::Command
+                                                     spec.triplet(),
+                                                     build_package_options,
+                                                     var_provider,
+                                                     std::move(*feature_dependencies),
```
These moves can get dropped, BuildPackageConfig takes const references to everything now.

cbezault commented on toolsrc/src/vcpkg/build.cpp:
```diff
@@ -713,6 +673,14 @@ namespace vcpkg::Build
                                       Hash::Algorithm::Sha1));
```
Above we add the feature_list to the abi file. `config.feature_list` is no longer stable and should be sorted before being added to the abi file.

JackBoosY commented:
This PR is what we need!

cbezault commented:
@vicroms, sigh, I probably introduced a bug in one of my last minute changes.

cbezault commented on toolsrc/src/vcpkg/dependencies.cpp:
```diff
@@ -33,17 +42,131 @@ namespace vcpkg::Dependencies
+                                    PackageSpec::from_name_and_triplet(dep.depend.name, m_spec.triplet())
+                                        .value_or_exit(VCPKG_LINE_INFO),
+                                    dep_feature);
+                            }
```
We need to add in the default `FeatureSpec` here. `core` is not in the feature list of `dep.depend.features`.

vicroms committed:
Fix issues with PR

vicroms committed:
Merge branch 'cleanup_supports' of https://github.com/Rastaban/vcpkg into pr/supports-expressions

cbezault commented:
/azp run

vicroms commented:
/azp run

cbezault committed:
Merge branch 'master' into cleanup_supports

vicroms commented on scripts/vcpkg_get_dep_info.cmake:
```diff
@@ -0,0 +1,17 @@
+    message("CMAKE_HOST_SYSTEM_NAME=${CMAKE_HOST_SYSTEM_NAME}")
+    message("CMAKE_HOST_SYSTEM_PROCESSOR=${CMAKE_HOST_SYSTEM_PROCESSOR}")
+    message("CMAKE_HOST_SYSTEM_VERSION=${CMAKE_HOST_SYSTEM_VERSION}")
+    message("CMAKE_HOST_SYSTEM=${CMAKE_HOST_SYSTEM}")
```
@cbezault 

Shouldn't `VCPKG_LIBRARY_LINKAGE` and `VCPKG_CRT_LINKAGE` be included here if we want to support the `static` qualifier?

cbezault replied:
Yup, oversight on my part.

cbezault committed:
Merge branch 'master' into cleanup_supports

vicroms committed:
[var_provider] Get library linkage variables from triplet

vicroms committed:
Merge branch 'cleanup_supports' of https://github.com/Rastaban/vcpkg into cleanup_supports

vicroms committed:
Fix compilation errors in tests

vicroms committed:
Add unimplemented functions

vicroms committed:
Fix unit tests part 1

vicroms committed:
Fix issue when buildtrees dir does not exist

vicroms committed:
Change binary output hash

cbezault commented:
/azp run

vicroms committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into cleanup_supports

cbezault committed:
Merge branch 'master' into cleanup_supports

vicroms commented on toolsrc/src/vcpkg/dependencies.cpp:
```diff
@@ -33,17 +42,136 @@ namespace vcpkg::Dependencies
+                {
+                    qualified_deps =
+                        &m_scfl.source_control_file->find_feature(feature).value_or_exit(VCPKG_LINE_INFO).depends;
+                }
```
This doesn't handle the `*` feature.

cbezault replied:
It shouldn't have to. The * feature should get expanded before this point. Do you have an example where this misbehaves?

vicroms replied:
It fails the basic feature tests, any install command should repro: `./vcpkg install cpprestsdk[*]`.

vicroms committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into cleanup_supports

vicroms committed:
Fix handling of * feature

vicroms committed:
Add core feature when using *

vicroms committed:
Do not add Default-Features when installing 'core'

vicroms commented:
Still failing upgrade plan tests.

cbezault commented:
I'll take a look when I can.

cbezault commented:
To summarize our offline conversation. My code doesn't handle the difference in the meaning of the core feature when specified on the command line and when written the the dependency list of a CONTROL file.

ras0219-msft committed:
Merge branch 'master' into cleanup_supports

cbezault committed:
Merge branch 'master' into cleanup_supports

cbezault commented:
/azp run

ras0219-msft committed:
Merge remote-tracking branch 'origin/master' into HEAD

cbezault committed:
Merge branch 'master' into cleanup_supports

ras0219-msft committed:
[vcpkg] WIP. 6 failing tests.

ras0219-msft committed:
[vcpkg] WIP. 1 failing tests.

ras0219-msft committed:
[vcpkg] WIP. 0 failing tests.

ras0219-msft committed:
[vcpkg] Removed 'remove_graph'. 0 failing tests.

ras0219-msft committed:
[vcpkg] Removed 'install_graph'. 0 failing tests.

ras0219-msft committed:
[vcpkg] Remove AnyAction; replace with ActionPlan

ras0219-msft committed:
Merge branch 'cleanup_supports' of https://github.com/Rastaban/vcpkg into HEAD

ras0219-msft committed:
[vcpkg] Minor cleanup.

ras0219-msft committed:
[vcpkg][z3][qt5-connectivity][qt5-purchasing] Improve error messages while parsing. Fix a few trivial port issues.

ras0219-msft committed:
[vcpkg] Work around ICE with MSVC v140

ras0219-msft committed:
[vcpkg] Add purge on fail to decompress for CI

ras0219-msft committed:
[vcpkg] Fix parsing of nested parentheses in qualifiers

ras0219-msft committed:
[vcpkg] Fix Linux builds (explicit qualification in declaration)

ras0219-msft committed:
Merge branch 'master' into cleanup_supports

ras0219-msft commented:
/azp run

ras0219-msft committed:
[vcpkg] Fix Build-Depends implying default features. Fix qualified dependencies regression.

ras0219-msft committed:
Merge remote-tracking branch 'origin/master' into HEAD

ras0219-msft committed:
[mmx] Add to skip list and full rebuild -- mmx causes problems by installing 'sched.h'

ras0219-msft committed:
Merge remote-tracking branch 'origin/master' into HEAD

ras0219-msft committed:
[libpqxx][mqtt-cpp] Prevent installing include/CMakeLists.txt

ras0219-msft committed:
[cppitertools] Fix installed include namespace (should be include/cppitertools)

ras0219-msft committed:
[libsoundio] Move headers into soundio/ subdirectory as per original cmake

ras0219-msft committed:
[ci.baseline] Temporarily skip charls due to conflict with dcmtk

ras0219-msft committed:
[vcpkg] Add restricted include files post build check -- bump global abi version

ras0219-msft committed:
[libsoundio] Hotfix stray line in portfile

ras0219-msft committed:
[vcpkg] Fix regression: CMake information was not being displayed for build-and-install actions

ras0219-msft committed:
[jsonnet] Fix installation of internal headers; use system nlohmann-json

ras0219-msft committed:
[grpc][upb] Teach grpc to use packaged upb. Add find_package(upb). Remove inappropriate upb features.

ras0219-msft committed:
[zfp] Move problematic 'include/bitstream.h' to 'include/zfp/bitstream.h'

ras0219-msft committed:
[x265] Bump control version to trigger rebuild after zfp conflict

ras0219-msft committed:
[akali] Disable parallel configure

ras0219-msft committed:
[dirent][dlfcn-win32][getopt-win32][pthreads] Grandfather into VCPKG_POLICY_ALLOW_RESTRICTED_HEADERS

ras0219-msft committed:
[ci.baseline] Update baseline for improved upb support

ras0219-msft committed:
[tgui] Disable parallel configure

ras0219-msft committed:
[libiconv] Enable VCPKG_POLICY_ALLOW_RESTRICTED_HEADERS

ras0219-msft committed:
[aws-sdk-cpp] Disable parallel configure

ras0219-msft committed:
[vcpkg] Implement policy VCPKG_POLICY_ALLOW_RESTRICTED_HEADERS

ras0219-msft committed:
[aws-sdk-cpp] Fix amount of escaping semicolons -- Note: I do not know the root cause requiring this change

ras0219-msft committed:
[libodb-sqlite] Fix configuring into source directory

ras0219-msft committed:
[gettext] Grandfather into VCPKG_POLICY_ALLOW_RESTRICTED_HEADERS

vicroms commented on toolsrc/src/vcpkg/postbuildlint.cpp:
```diff
@@ -98,6 +98,90 @@ namespace vcpkg::PostBuildLint
+            "parser.h",
+            "lexer.h",
+        };
+        static constexpr Span<const StringLiteral> restricted_lists[] = {
```
Shouldn't we add C++ Standard Library header names as well? At least one library (`sfsexp`) is known to install `cstring.h` directly into `include/`.

ras0219-msft committed:
[libodb] DISABLE_PARALLEL_CONFIGURE

ras0219-msft committed:
[vcpkg] Add 'config.h' and 'local.h' to restricted header list

ras0219-msft committed:
Merge remote-tracking branch 'origin/master' into HEAD

ras0219-msft committed:
[mcpp] Remove unused and problematic include 'config.h' from installed files

ras0219-msft committed:
[teemo] Move installed headers into subdirectory to prevent conflicts with x265

ras0219-msft committed:
[ci.baseline] Update current OSX. Skip libmesh on all platforms due to heavy conflicts.

ras0219-msft committed:
[vcpkg] Add 'slice.h' as a restricted header

ras0219-msft committed:
[osg] Improve accuracy of dependencies (disable some, add some to Depends)

ras0219-msft committed:
[vcpkg] Skip invoking a subprocess for 0 specs in load_tag_vars

ras0219-msft committed:
[ci.baseline] Skip mongo-c-driver on osx due to flakiness

ras0219-msft committed:
[teemo] Fix incorrect include file read

ras0219-msft committed:
[osg] Fix dependency typo: glut -> freeglut

ras0219-msft committed:
Merge remote-tracking branch 'origin/master' into HEAD

ras0219-msft committed:
[vcpkg] Recover some lost performance with the addition of vcpkg_get_tags.

A huge performance cost was loading the triplet files over and over; instead, we splice the sources into a macro and load it once, then just call that macro for each port.
Remove use of hashing because we aren't cross-process-safe anyway (global static will do instead).

dan-shaw commented:
This looks like it has passed the CI. Before merging this, can we add more documentation as to what was changed (such as adding a new build policy)?

ras0219-msft commented:
Also, @winsoft666, as part of this PR I discovered that `teemo` installing `include/slice.h` caused x264 to fail to build (and it's easy to see that `#include "slice.h"` is likely to conflict with other libraries).

I resolved this by moving the includes into `include/teemo/` -- CMake projects should be unaffected and can continue using the same include paths, but MSBuild projects will now need to use `#include <teemo/teemo.h>`. I'd strongly recommend considering making this change upstream and having all users always include `teemo/teemo.h`.

ras0219-msft committed:
[vcpkg] Change Supports atom 'windows' to include UWP. Improve Supports field documentation.

ras0219-msft committed:
[vcpkg] Add docs for VCPKG_ENV_PASSTHROUGH and VCPKG_DEP_INFO_OVERRIDE_VARS

cbezault commented on docs/maintainers/control-files.md:
```diff
@@ -96,38 +96,55 @@ Vcpkg does not distinguish between build-only dependencies and runtime dependenc
 
 #### Supports
-Expression that evalutates to true when the port is supported for a given triplet.  This field is currently only used in the CI testing to skip ports.
+Expression that evalutates to true when the port is expected to build successfully for a triplet.
```
Evaluates

vicroms committed:
Fix typo

vicroms reviewed (approved):
Let's Get This Merged

cbezault commented:
bless

JackBoosY commented:
With this PR, I think we can further improve the CI test to add feature tests.
